File: packages/php-sdk/src/Resources/Console.php
<?php

namespace Wusul\Resources;

use Wusul\HttpClient;

/**
 * Console resource for managing card templates (Enterprise only)
 */
class Console
{
    private HttpClient $http;

    /**
     * Creates a new Console resource instance
     *
     * @param HttpClient $httpClient The HTTP client
     */
    public function __construct(HttpClient $httpClient)
    {
        $this->http = $httpClient;
    }

    /**
     * Create a new card template
     *
     * @param array $params Parameters for creating the card template
     *                      - name (required): Template name
     *                      - platform (required): 'apple' or 'google'
     *                      - useCase (required): 'employee_badge' or 'hotel'
     *                      - protocol (required for Apple): 'desfire' or 'seos'
     *                      - allowOnMultipleDevices: Allow on multiple devices
     *                      - watchCount: Number of watches allowed
     *                      - iphoneCount: Number of iPhones allowed
     *                      - design: Design configuration object
     *                      - supportInfo: Support information object
     *                      - metadata: Custom JSON object
     * @return array The created card template
     * @throws \Exception
     */
    public function createTemplate(array $params): array
    {
        return $this->http->post('/v1/console/card-templates', $params);
    }

    /**
     * Read a card template
     *
     * @param string $cardTemplateId The ID of the card template to read
     * @return array The card template
     * @throws \Exception
     */
    public function readTemplate(string $cardTemplateId): array
    {
        return $this->http->get(
            "/v1/console/card-templates/{$cardTemplateId}",
            ['id' => $cardTemplateId]
        );
    }

    /**
     * Update a card template
     *
     * @param array $params Parameters for updating the card template
     *                      - cardTemplateId (required): The template ID
     *                      - name: Updated name
     *                      - allowOnMultipleDevices: Updated multi-device setting
     *                      - watchCount: Updated watch count
     *                      - iphoneCount: Updated iPhone count
     *                      - supportInfo: Updated support info
     *                      - metadata: Updated metadata
     * @return array The updated card template
     * @throws \Exception
     */
    public function updateTemplate(array $params): array
    {
        $cardTemplateId = $params['cardTemplateId'] ?? null;
        if (!$cardTemplateId) {
            throw new \InvalidArgumentException('cardTemplateId is required');
        }

        unset($params['cardTemplateId']);
        return $this->http->patch("/v1/console/card-templates/{$cardTemplateId}", $params);
    }

    /**
     * Publish a card template
     *
     * @param string $cardTemplateId The ID of the card template to publish
     * @return array Success response
     * @throws \Exception
     */
    public function publishTemplate(string $cardTemplateId): array
    {
        return $this->http->post("/v1/console/card-templates/{$cardTemplateId}/publish");
    }

    /**
     * Read event logs for a card template
     *
     * @param array $params Parameters for filtering event logs
     *                      - cardTemplateId (required): The template ID
     *                      - filters: Optional filters object
     *                          - device: 'mobile' or 'watch'
     *                          - startDate: ISO8601 timestamp
     *                          - endDate: ISO8601 timestamp
     *                          - eventType: 'issue', 'install', 'update', 'suspend', 'resume', or 'unlink'
     * @return array Array of event log entries
     * @throws \Exception
     */
    public function eventLog(array $params): array
    {
        $cardTemplateId = $params['cardTemplateId'] ?? null;
        if (!$cardTemplateId) {
            throw new \InvalidArgumentException('cardTemplateId is required');
        }

        $sigPayload = ['id' => $cardTemplateId];
        $filters = $params['filters'] ?? null;

        if ($filters) {
            if (isset($filters['device'])) {
                $sigPayload['device'] = $filters['device'];
            }
            if (isset($filters['startDate'])) {
                $sigPayload['start_date'] = $filters['startDate'];
            }
            if (isset($filters['endDate'])) {
                $sigPayload['end_date'] = $filters['endDate'];
            }
            if (isset($filters['eventType'])) {
                $sigPayload['event_type'] = $filters['eventType'];
            }
        }

        return $this->http->get("/v1/console/card-templates/{$cardTemplateId}/logs", $sigPayload);
    }

    /**
     * iOS Preflight - Get provisioning identifiers for iOS
     * (Enterprise only)
     *
     * @param array $params Parameters
     *                      - cardTemplateId (required): The template ID
     *                      - accessPassExId (required): The access pass ex_id
     * @return array Provisioning identifiers
     * @throws \Exception
     */
    public function iosPreflight(array $params): array
    {
        $cardTemplateId = $params['cardTemplateId'] ?? null;
        $accessPassExId = $params['accessPassExId'] ?? null;

        if (!$cardTemplateId || !$accessPassExId) {
            throw new \InvalidArgumentException('cardTemplateId and accessPassExId are required');
        }

        return $this->http->get(
            "/v1/console/card-templates/{$cardTemplateId}/ios-preflight/{$accessPassExId}",
            ['id' => $accessPassExId]
        );
    }
}

File: packages/ruby-sdk/lib/wusul/resources/console.rb
module Wusul
  module Resources
    # Console resource for managing card templates (Enterprise only)
    class Console
      # Initialize the Console resource
      #
      # @param http_client [HttpClient] The HTTP client instance
      def initialize(http_client)
        @http = http_client
      end

      # Create a new card template (Enterprise only)
      #
      # @param params [Hash] Parameters for creating the card template
      # @option params [String] :name Display name for the template
      # @option params [String] :platform Must be 'apple' or 'google'
      # @option params [String] :use_case Must be 'employee_badge' or 'hotel'
      # @option params [String] :protocol Must be 'desfire' or 'seos'
      # @option params [Boolean, nil] :allow_on_multiple_devices Allow on multiple devices
      # @option params [Integer, nil] :watch_count Number of watches (1-5)
      # @option params [Integer, nil] :iphone_count Number of iPhones (1-5)
      # @option params [Hash, nil] :design Template design settings
      # @option params [Hash, nil] :support_info Support information
      # @option params [Hash, nil] :metadata Custom metadata
      # @return [Hash] The created card template
      def create_template(params)
        @http.post('/v1/console/templates', params)
      end

      # Update an existing card template (Enterprise only)
      #
      # @param params [Hash] Parameters for updating the card template
      # @option params [String] :card_template_id The template ID to update
      # @option params [String, nil] :name Updated display name
      # @option params [Boolean, nil] :allow_on_multiple_devices Allow on multiple devices
      # @option params [Integer, nil] :watch_count Number of watches (1-5)
      # @option params [Integer, nil] :iphone_count Number of iPhones (1-5)
      # @option params [Hash, nil] :support_info Updated support information
      # @option params [Hash, nil] :metadata Updated custom metadata
      # @return [Hash] The updated card template
      def update_template(params)
        template_id = params.delete(:card_template_id)
        raise ArgumentError, 'card_template_id is required' unless template_id

        @http.patch("/v1/console/templates/#{template_id}", params)
      end

      # Read card template information (Enterprise only)
      #
      # @param card_template_id [String] Unique identifier for the card template
      # @return [Hash] The card template details
      def read_template(card_template_id)
        raise ArgumentError, 'card_template_id is required' unless card_template_id

        sig_payload = { template_id: card_template_id }
        @http.get("/v1/console/templates/#{card_template_id}", sig_payload)
      end

      # Read event log for a card template (Enterprise only)
      #
      # @param params [Hash] Parameters for reading event log
      # @option params [String] :card_template_id Template ID to query
      # @option params [Hash, nil] :filters Optional filters for events
      # @return [Array<Hash>] Array of event log entries
      def event_log(params)
        template_id = params[:card_template_id]
        raise ArgumentError, 'card_template_id is required' unless template_id

        sig_payload = { template_id: template_id }
        sig_payload[:filters] = params[:filters] if params[:filters]

        @http.get("/v1/console/templates/#{template_id}/events", sig_payload)
      end

      # Get iOS provisioning identifiers (Enterprise only)
      #
      # @param params [Hash] Parameters for iOS preflight
      # @option params [String] :card_template_id The card template ID
      # @option params [String] :access_pass_ex_id The access pass ex_id
      # @return [Hash] iOS provisioning identifiers
      def ios_preflight(params)
        template_id = params[:card_template_id]
        pass_id = params[:access_pass_ex_id]

        raise ArgumentError, 'card_template_id is required' unless template_id
        raise ArgumentError, 'access_pass_ex_id is required' unless pass_id

        sig_payload = {
          card_template_id: template_id,
          access_pass_ex_id: pass_id
        }

        @http.get('/v1/console/ios-preflight', sig_payload)
      end
    end
  end
end
